### DISCUSSION TITLE
Loader configuration in monorepo [Module parse failed: Unexpected token]

### DISCUSSION THREAD

**Discussion by Poolshark** [webpack/webpack#15753] [Q&A] [answered]
I'm trying to set up a centralised package from which I can bundle/start various other projects and apps within my monorepo.

The monorepo has the following structure:
```
config/
  webpack/
    config/
      apps.webpack.config.ts
    webpack.config.dev.ts
  package.json
frontend/
  apps/
    app1/
      src/
```

where `apps.webpack.ts` holds basic configurations like _fallbacks_, _aliases_, ... and **rules**, where the `babel-loader` is defined and configured. The webpack.config.dev.ts is the configuration file which includes the definitions from `apps.webpack.config.ts`, so

```ts
// apps.webpack.ts

...

export const rules: NonNullable<WebpackConfiguration["module"]>["rules"] = [
  { test: /\.(svg|ico|png|gif|jpg|jpeg)$/, type: "asset/ressource" },
  {
    test: /\.(ts|tsx|js|jsx)?$/,
    include: [resolve(__dirname, "../../../frontend")],    // I'm not sure if here...
    use: {
      loader: require.resolve("babel-loader"),
      options: {
        include: [resolve(__dirname, "../../../frontend")], // ...or here?
        generatorOpts: { compact: false },
        presets: ["@babel/preset-env", ["@babel/preset-react", { runtime: "automatic" }], "@babel/preset-typescript"],
        plugins: [
          ["relay", { artifactDirectory: "../../../frontend/__generated__", eagerESModules: true }],
          ["@babel/plugin-transform-runtime"],
        ],
      },
    },
  },
];
```
and
```ts
// webpack.config.dev.ts

import { rules } from "../config/apps.webpack.ts"

const configurationBuilder = (env: WebpackEnv, argv: any): Config => {
  return {
    mode: env.mode ? env.mode : "development",
    output: {
      publicPath: "/",
    },
    entry: resolve(__dirname, `dirToApp/src/index.tsx`),
    ...
    module: {
      rules: rules
  };
};

export default configurationBuilder;
```

I now want to run the React application `app1` in my `frontend/` directory via my yarn script
```json
// package.json

...
"scripts": {
  "start:app1": "webpack serve --config webpack/webpack.config.dev.ts --env=app=app1 --env=mode=development"
 }
```

It seems like that everything is going well but unfortunately I get the error

```
ERROR in ../frontend/apps/app1/src/index.tsx 10:4
Module parse failed: Unexpected token (10:4)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|   const root = createRoot(container);
|   root.render(
>     <StrictMode>
|       <App />
|     </StrictMode>
```

I guess this has something to do with `include`, right?

**Comment by vankop** [comment]
could be 2 problems:
-  `include` part of rule ( you can test this just removing `include` part )
-  babel configuration

**Reply by Poolshark** [reply]
Hey! Thank's for the quick answer. I get the same error when I remove the `include` part. Tried it
- without `include`
- with `include` under `test` only (this is where TS gives the type hint for it)
- with `include` in `options` section only
- with `include` in both places

So I gess I have a problem with the babel configuration? To clarify: the config works if I move the `webpack.config.dev.ts` inside the `apps1/` folder!

**Reply by vankop** [reply]
please create reproducible repo, otherwise hard to guess where is a problem

**Comment by Poolshark** [comment, marked as answer]
Found the solution! 🙌

In my `tsconfig.json` in my `config` project (the one I run Webpack from), I had path aliases defined which I imported into the `webpack.config.dev.ts`. Obviously Webpack does not know about these path aliases unless I specifically define those in the `alias` section.

So my `tsconfig.json` is
```json
{
  "compilerOptions": {
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true,
    "noEmit": true,
    "incremental": true,
    "downlevelIteration": true,
    "esModuleInterop": true,
    "rootDir": "../",
    "baseUrl": "."
    "paths": {
      "webpack-config": ["./config"]
    }
  },
  "include": ["./typescript", "./webpack", "../frontend"]
}
```

Then, in my `apps.webpack.config.ts` I have a section for all the aliases which I use for my apps/webpack
```ts
export const aliases: NonNullable<WebpackConfiguration["resolve"]>["alias"] = {
  webpack-config: resolve(__dirname),
  ...
};
```

Hope this helps!
Thank you @vankop for the very quick response! 🙌
